### ISSUE TITLE
Error with Warnings and Ntuple Writeout

### ISSUE THREAD

**Issue by GNiendorf** [SegmentLinking/TrackLooper#354]
If the warning printouts are enabled, and you run for example `./bin/sdl -n 175 -s 8 -w 1`  which has the ntuple writout enabled, this will throw the following error. Found by @ariostas from PR #351:

```
ERROR
 hittype: 0
 trk.pix_simHitIdx().size(): 162652
 trk.ph2_simHitIdx().size(): 213716
162652 0
206907 0
terminate called after throwing an instance of 'std::out_of_range'
  what():  vector::_M_range_check: __n (which is 206907) >= this->size() (which is 162652)
```

**Comment by ariostas** [comment]
The issue happens because of this line, and it only happens with CUDA. I'm still not sure why this happens.

https://github.com/SegmentLinking/TrackLooper/blob/17e0885cd71e988d57bfe619bedb07ded14735f9/SDL/Segment.h#L738

**Comment by GNiendorf** [comment]
@ariostas If you serialize that kernel in CUDA then it runs fine. Very strange.

**Comment by ariostas** [comment]
So it turns out that what's breaking is the atomic addition in this line.
https://github.com/SegmentLinking/TrackLooper/blob/17e0885cd71e988d57bfe619bedb07ded14735f9/SDL/Segment.h#L742

I still don't fully understand why this happens, but my guess is that since `printf` probably uses some kind of atomic operation to add stuff to the buffer that is printed out after the kernel finishes, then it ends up using some kind of atomic operation for the branching of that `if` statement, and ends up conflicting with the atomic addition. So to make sure that this doesn't happen I'll try to move the warnings to the end of kernels and far away from atomic operations.

**Comment by ariostas** [comment]
I wrote minimal examples that have this bug using Alpaka and with pure CUDA. They can be found [here](https://gist.github.com/ariostas/255a01d1c3ab1dd9b85142e1bcf180e4).

**Comment by slava77** [comment]
> I wrote minimal examples that have this bug using Alpaka and with pure CUDA. They can be found [here](https://gist.github.com/ariostas/255a01d1c3ab1dd9b85142e1bcf180e4).

What's the expected symptom of the bug in this example?

**Comment by ariostas** [comment]
> What's the expected symptom of the bug in this example?

The two kernel calls produce different outputs. The sum in the version that is compiled with the `printf` is wrong.

**Comment by slava77** [comment]
> > What's the expected symptom of the bug in this example?
> 
> The two kernel calls produce different outputs. The sum in the version that is compiled with the `printf` is wrong.

is there a pattern? e.g. as if `n2=0` is used

**Comment by ariostas** [comment]
We've discussed what happened during the meetings, but I'll write it here so we can keep it on record.

I reached out to NVIDIA and told them about a minimal example of the bug that I posted [here](https://gist.github.com/ariostas/b159a4aa9cfb8582043765d23c9e1117). They were able to reproduce the bug. After some back and forth, they said that the reason was that there was a buggy JIT compilation that was causing the issue. For newer cards it could be fixed by specifically compiling for that compute capability, but for older cards all JIT optimizations needed to be disabled. The issue is now fixed in drivers >= 550.54.14.

For our purposes, cgpu-1 is using the latest driver, lnx4555 has a newer card and we are compiling specifically for that compute capability, but lnx7188 has a driver with the bug, an older card, and we are not disabling JIT optimizations, so the bug is still present there. I think it's probably fine to close this issue, but I'll let someone else make that decision.
